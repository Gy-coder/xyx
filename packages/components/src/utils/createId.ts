let id = 0;

export default function createId(){
    return id++
}