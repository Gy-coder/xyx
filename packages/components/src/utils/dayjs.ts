type DayUnit = "year"
    | "month"
    | "day"
    | "hour"
    | "minute"
    | "second"
    | "millisecond";

export default class Dayjs {
    public date: Date

    constructor(date?: number | string | Date) {
        this.date = date ? new Date(date) : new Date()
    }

    get year() {
        return this.date.getFullYear()
    }

    get month() {
        return this.date.getMonth() + 1
    }

    get week() {
        return this.date.getDay()
    }

    get day() {
        return this.date.getDate()
    }

    get hour() {
        return this.date.getHours()
    }

    get minute() {
        return this.date.getMinutes()
    }

    get second() {
        return this.date.getSeconds()
    }

    get millisecond() {
        return this.date.getMilliseconds()
    }

    get raw() {
        return this.date
    }

    public firstDayOfMonth() {
        return new Dayjs(
            new Date(
                this.year,
                this.month - 1,
                1,
                0,
                0,
                0)
        )
    }

    public add(amount: number, unit: DayUnit) {
        let date = new Date(this.date.getTime());
        switch (unit) {
            case 'year':
                const currentDate = date.getDate()
                date.setDate(1)
                date.setFullYear(date.getFullYear() + amount)
                const targetDate = new Date(
                    date.getFullYear(),
                    date.getMonth() + 1,
                    0,
                    0,
                    0,
                    0,
                ).getDate()
                date.setDate(Math.min(currentDate, targetDate))
                break;
            case 'month':
                const d = date.getDate()
                date.setDate(1)
                date.setMonth(date.getMonth() + amount);
                const d2 = new Date(date.getFullYear(), date.getMonth() + 1, 0, 0, 0, 0).getDate()
                date.setDate(Math.min(d, d2))
                break;
            case 'day':
                date.setDate(date.getDate() + amount);
                break;
            case 'hour':
                date.setHours(date.getHours() + amount);
                break;
            case 'minute':
                date.setMinutes(date.getMinutes() + amount);
                break;
            case 'second':
                date.setSeconds(date.getSeconds() + amount);
                break;
            case 'millisecond':
                date.setMilliseconds(date.getMilliseconds() + amount);
                break;

            default:
                throw new Error('Time.add: unknown unit');
        }
        return new Dayjs(date)
    }

    public set(amount: number,unit: DayUnit){
        const date = new Date(this.date.getTime())
        switch (unit){
            case "year":
                date.setFullYear(amount)
                break
            case "month":
                date.setMonth(amount - 1)
                break
            case "day":
                date.setDate(amount)
                break
            case "hour":
                date.setHours(amount)
                break
            case "second":
                date.setSeconds(amount)
                break
            case "millisecond":
                date.setMilliseconds(amount)
                break
            default:
                throw new Error("Time.add: unknown unit")
        }
        return new Dayjs(date)
    }

    isSameDay(day: Dayjs){
        return day.year === this.year && day.month === this.month && day.day === this.day
    }

    isSameMonth(day: Dayjs){
        return day.year === this.year && day.month === this.month
    }

    //@ts-ignore
    format(pattern?: string = 'YYYY-MM-DD') {
        // 目前支持的格式有 YYYY MM DD HH mm ss SSS
        return pattern.replace(/YYYY/g, this.year.toString())
            .replace(/MM/, this.month.toString().padStart(2, '0'))
            .replace(/DD/, this.day.toString().padStart(2, '0'))
            .replace(/HH/, this.hour.toString().padStart(2, '0'))
            .replace(/mm/, this.minute.toString().padStart(2, '0'))
            .replace(/ss/, this.second.toString().padStart(2, '0'))
            .replace(/SSS/, this.millisecond.toString().padStart(3, '0'))
    }
}
