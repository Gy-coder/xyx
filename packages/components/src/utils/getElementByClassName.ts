export function getElementByClassName(container = document, classname: string) {
  return container.getElementsByClassName(classname)[0];
}
