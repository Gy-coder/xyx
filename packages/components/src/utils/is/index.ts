export { default as isUndefined } from "./isUndefined";
export { default as isFunction } from "./isFunction";
