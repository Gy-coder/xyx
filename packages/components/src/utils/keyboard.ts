export const keyboard = {
  Enter: "Enter",
};
