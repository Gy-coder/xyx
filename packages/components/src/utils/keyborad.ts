export const keyborad = {
  Enter: "Enter",
};
