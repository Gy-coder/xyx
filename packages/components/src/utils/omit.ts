export default function omit<T = any>(obj: Record<string, T>, paths: string[]) {
  const res: Record<string, T> = { ...obj },
    set = new Set(paths);
  for (let key in obj) {
    if (set.has(key)) delete res[key];
  }
  return res;
}
