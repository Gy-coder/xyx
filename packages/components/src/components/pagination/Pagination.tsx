import React, {forwardRef, ForwardRefRenderFunction} from "react";
import classnames from "classnames";
import {PaginationProps} from "./interface";
import usePager from "./usePager";
import InternalIcon from "../_interal/internal_icon/InternalIcon";
import Pager from "./Pager";
import Jumper from "./PaginationJumper";
import SimplePager from "./SimplePager";
import {isUndefined} from "../../utils/is";
import "./index.scss";

const Pagination: ForwardRefRenderFunction<any, PaginationProps> = (
    props,
    ref
) => {
    const {
        defaultCurrent,
        current,
        onChange,
        total = 1,
        pageSize = 10,
        pagerCount = 7,
        showQuickJumper = false,
        simple = false,
        hideOnSinglePage = false,
        mini = false
    } = props;

    if (!isUndefined(current) && isUndefined(onChange))
        console.error("current property must be with onChange")

    const {pageIndex, prevPage, nextPage, totalPage, jumpToPage} = usePager({
        defaultCurrent,
        current,
        onChange,
        total,
        pageSize,
    });

    const renderPager = simple ? (
        <SimplePager
            pageIndex={pageIndex}
            onChange={jumpToPage}
            totalPage={totalPage}
        />
    ) : (
        <Pager
            pageIndex={pageIndex}
            onChange={jumpToPage}
            totalPage={totalPage}
            pagerCount={pagerCount}
            mini={mini}
        />
    );

    const paginationClasses = classnames("g-pagination",{
        "g-pagination-mini": mini
    })
    const buttonClasses = (pos: "prev" | "next", index: number) =>
        classnames(`g-pagination-button`, `g-pagination-${pos}-button`, {
            [`g-pagination-button-disabled`]: pageIndex === index,
            [`g-pagination-button-mini`]: simple || mini,
        });

    const hidePagination = hideOnSinglePage && totalPage === 1;

    return !hidePagination ? (
        <div className={paginationClasses} ref={ref}>
            <button className={buttonClasses("prev", 1)} onClick={prevPage}>
                <InternalIcon name="icon-arrow-left"/>
            </button>
            {renderPager}
            <button className={buttonClasses("next", totalPage)} onClick={nextPage}>
                <InternalIcon name="icon-arrow-right"/>
            </button>
            {showQuickJumper && !simple && <Jumper onChange={jumpToPage} mini={mini}/>}
        </div>
    ) : null;
};

export default forwardRef(Pagination);
