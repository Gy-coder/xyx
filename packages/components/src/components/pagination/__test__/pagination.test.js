import {fireEvent, render, screen} from "@testing-library/react";
import "@testing-library/jest-dom";
import {getElementByClassName} from "../../../utils/getElementByClassName";
import Pagination from "../Pagination";
import {useState} from "react";
import {keyboard} from "../../../utils/keyboard";

describe("test Pagination component", () => {
    it("should render correctly", () => {
        const {container, asFragment} = render(<Pagination total={100}/>);
        const el = getElementByClassName(container, "g-pagination");
        expect(el).toBeVisible();
        expect(asFragment()).toMatchSnapshot()
    });
    it("can click prev at 1", () => {
        const jestFn = jest.fn();
        let {container} = render(<Pagination total={100} onChange={jestFn}/>);
        let leftButton = getElementByClassName(
            container,
            "g-pagination-prev-button"
        );
        expect(leftButton).toBeVisible();
        fireEvent.click(leftButton);
        expect(jestFn).toBeCalledWith(1);
    });
    it("can click prev at 2", () => {
        const jestFn = jest.fn();
        const {container} = render(
            <Pagination total={100} onChange={jestFn} defaultCurrent={2}/>
        );
        let leftButton = getElementByClassName(
            container,
            "g-pagination-prev-button"
        );
        expect(leftButton).toBeVisible();
        fireEvent.click(leftButton);
        expect(jestFn).toBeCalledWith(1);
    });
    it("can click next at 1", () => {
        const jestFn = jest.fn();
        const {container} = render(<Pagination onChange={jestFn} total={100}/>);
        const rightButton = getElementByClassName(
            container,
            "g-pagination-next-button"
        );
        expect(rightButton).toBeVisible();
        fireEvent.click(rightButton);
        expect(jestFn).toBeCalledWith(2);
    });
    it("can click next at totalPage", () => {
        const jestFn = jest.fn();
        const {container} = render(
            <Pagination onChange={jestFn} total={100} defaultCurrent={10}/>
        );
        const rightButton = getElementByClassName(
            container,
            "g-pagination-next-button"
        );
        expect(rightButton).toBeVisible();
        fireEvent.click(rightButton);
        expect(jestFn).toBeCalledWith(10);
    });
    it("controlled component", () => {
        const jestFn = jest.fn()
        let v, setV, handleChange
        const Demo = () => {
            [v, setV] = useState(1)
            handleChange = (newCurrent) => {
                jestFn(newCurrent)
                setV(newCurrent)
            }
            return <>
                <Pagination total={1000} current={v} onChange={handleChange}/>
            </>
        }
        const {container} = render(<Demo/>)
        const six = screen.getByText("6")
        fireEvent.click(six)
        expect(jestFn).toBeCalledWith(6)
        expect(v).toBe(6)
    })
    it("hideOnSinglePage", () => {
        const {container, asFragment} = render(<Pagination total={1} hideOnSinglePage/>)
        const el = getElementByClassName(container, "g-pagination")
        expect(el).toBeFalsy()
        expect(asFragment()).toMatchSnapshot()
    })
    xit("test simple", () => {
        const jestFn = jest.fn()
        const {container, asFragment} = render(<Pagination total={100} simple onChange={jestFn}/>)
        const el = getElementByClassName(container, "g-input-origin")
        fireEvent.change(el, {target: {value: "6"}})
        fireEvent.keyPress(el, {code: keyboard.Enter})
        expect(jestFn).toBeCalledWith(6)
        expect(asFragment()).toMatchSnapshot()
    })
    it("test mini",()=> {
        const {container} = render(<Pagination total={100} mini />)
        const el = getElementByClassName(container,'g-pagination')
        expect(el).toHaveClass('g-pagination-mini')
    })
});
