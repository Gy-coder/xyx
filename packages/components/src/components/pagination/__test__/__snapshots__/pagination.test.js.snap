// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Pagination component hideOnSinglePage 1`] = `<DocumentFragment />`;

exports[`test Pagination component should render correctly 1`] = `
<DocumentFragment>
  <div
    class="g-pagination"
  >
    <button
      class="g-pagination-button g-pagination-prev-button g-pagination-button-disabled"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="#icon-arrow-left"
        />
      </svg>
    </button>
    <ul
      class="g-pagination-pager"
    >
      <li
        class="g-pagination-pager-item g-pagination-pager-item-active"
      >
        1
      </li>
      <li
        class="g-pagination-pager-item"
      >
        2
      </li>
      <li
        class="g-pagination-pager-item"
      >
        3
      </li>
      <li
        class="g-pagination-pager-item"
      >
        4
      </li>
      <li
        class="g-pagination-pager-item"
      >
        5
      </li>
      <li
        class="g-pagination-pager-item"
      >
        6
      </li>
      <li
        class="g-pagination-pager-item"
      >
        <svg
          class="icon"
        >
          <use
            xlink:href="#icon-ellipsis"
          />
        </svg>
      </li>
      <li
        class="g-pagination-pager-item"
      >
        10
      </li>
    </ul>
    <button
      class="g-pagination-button g-pagination-next-button"
    >
      <svg
        class="icon"
      >
        <use
          xlink:href="#icon-arrow-right"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
